src/core/memory/freezer.cpp

```diff
@@ -0,0 +1,186 @@
+void Freezer::SetActive(bool active) {
+    if (!this->active.exchange(active)) {
+        FillEntryReads();
+        core_timing.ScheduleEvent(MEMORY_FREEZER_TICKS, event);
+        LOG_DEBUG(Common_Memory, "Memory freezer activated!");
+    } else {
+        LOG_DEBUG(Common_Memory, "Memory freezer deactivated!");
+    }
+}
+
+bool Freezer::IsActive() const {
+    return active.load();
+}
+
+void Freezer::Clear() {
+    std::lock_guard<std::recursive_mutex> lock(entries_mutex);
+
+    LOG_DEBUG(Common_Memory, "Clearing all frozen memory values.");
+
+    entries.clear();
+}
+
+u64 Freezer::Freeze(VAddr address, u8 width) {
+    std::lock_guard<std::recursive_mutex> lock(entries_mutex);
+
+    const auto current_value = MemoryReadWidth(width, address);
+    entries.push_back({address, width, current_value});
+
+    LOG_DEBUG(Common_Memory,
+              "Freezing memory for address={:016X}, width={:02X}, current_value={:016X}", address,
+              width, current_value);
+
+    return current_value;
+}
+
+void Freezer::Unfreeze(VAddr address) {
+    std::lock_guard<std::recursive_mutex> lock(entries_mutex);
+
+    LOG_DEBUG(Common_Memory, "Unfreezing memory for address={:016X}", address);
+
+    entries.erase(
+        std::remove_if(entries.begin(), entries.end(),
+                       [&address](const Entry& entry) { return entry.address == address; }),
+        entries.end());
+}
+
+bool Freezer::IsFrozen(VAddr address) {
+    std::lock_guard<std::recursive_mutex> lock(entries_mutex);
+
+    return std::find_if(entries.begin(), entries.end(), [&address](const Entry& entry) {
+               return entry.address == address;
+           }) != entries.end();
+}
+
+void Freezer::SetFrozenValue(VAddr address, u64 value) {
+    std::lock_guard<std::recursive_mutex> lock(entries_mutex);
+
+    const auto iter = std::find_if(entries.begin(), entries.end(), [&address](const Entry& entry) {
+        return entry.address == address;
+    });
+
+    if (iter == entries.end()) {
+        LOG_ERROR(Common_Memory,
+                  "Tried to set freeze value for address={:016X} that is not frozen!", address);
+        return;
+    }
+
+    LOG_DEBUG(Common_Memory,
+              "Manually overridden freeze value for address={:016X}, width={:02X} to value={:016X}",
+              iter->address, iter->width, value);
+    iter->value = value;
+}
+
+std::optional<Freezer::Entry> Freezer::GetEntry(VAddr address) {
+    std::lock_guard<std::recursive_mutex> lock(entries_mutex);
+
+    const auto iter = std::find_if(entries.begin(), entries.end(), [&address](const Entry& entry) {
+        return entry.address == address;
+    });
+
+    if (iter == entries.end()) {
+        return std::nullopt;
+    }
+
+    return *iter;
+}
+
+std::vector<Freezer::Entry> Freezer::GetEntries() {
+    std::lock_guard<std::recursive_mutex> lock(entries_mutex);
+
+    return entries;
+}
+
+void Freezer::FrameCallback(u64 userdata, s64 cycles_late) {
+    if (!active.load()) {
```

Comment:
```suggestion
    if (!IsActive()) {
```

Also are you certain this should be appearing before the mutex being locked?

Reply:
Yes, checking activity/stopping the callback if there is none doesn't read or write the entries vector.